### PR TITLE
Add Oauth2

### PR DESCRIPTION
**Pull request by mboylevt**
First attempt at a PR to add Oauth2 authentication steps + materials endpoint w/ bearer token.  Example included

### CODE REVIEW

File: swApiClient.cpp [hunk 1]
```diff
@@ -13,34 +30,80 @@ swApiClient Constructor
...
+  string readBuffer;
+
+  // prepare auth payload
+  char *postData = (char*)malloc(250);
+  strcpy(postData, "grant_type=client_credentials&client_id=");
+  strcat(postData, this->consumerKey.c_str());
+  strcat(postData, "&client_secret=");
+  strcat(postData, this->consumerSecret.c_str());
+
+  curl = curl_easy_init();
```
Comment (andreybrokeit) [inline swApiClient.cpp:58]: cool kids just do `CURL *curl = curl_easy_init();` :) but those same cool kids also participate in the worst readable C code I think

**Comment by nicolewagoner** [comment]
Oauth2! 👍

File: Makefile [hunk 2]
```diff
@@ -11,11 +11,11 @@ MKDIR = mkdir -p
 
 all: MAKEDIR $(TARGET)
 
-MAKEDIR: 
+MAKEDIR:
 	$(MKDIR) bin
 
 $(TARGET): main.cpp
-	$(CC) $(CLFAGS) -o bin/$(TARGET) main.cpp swApiClient.cpp -lcurl
```
Comment (d33psky) [inline Makefile:18]: nice typo :-D

File: main.cpp [hunk 3]
```diff
@@ -1,31 +1,10 @@
...
-    curl_easy_setopt(curl, CURLOPT_URL, "http://api.shapeways.com");
-    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
-    res = curl_easy_perform(curl);
-    curl_easy_cleanup(curl);
-
-    std::cout << readBuffer << std::endl;
-  }
-  return 0;
+  swApiClient * client = new swApiClient("CLIENT_KEY", "CLIENT_SECRET");
```
Comment (d33psky) [inline main.cpp]: are those client_key and _secret the oauth2 variants of the Consumer Key and Consumer Secret, or the access token ? I think the Consumer Key and Secret. Let's use those names then.
Also lateron you use style 
swApiClient *client =

File: swApiClient.cpp [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
 
-/* static member init */
-std::string swApiClient::BASE_URL = "http://api.shapeways.com";
-std::string swApiClient::API_VERSION = "v1";
+/* Helper Functions */
+struct curl_slist * swApiClient::SetAuthHeader() {
+  char *headerStr = (char*)malloc(250);
+  strcpy(headerStr, "Authorization: Bearer ");
+  strcat(headerStr, this->accessToken.c_str());
+  cout << headerStr << endl;
```
Comment (d33psky) [inline swApiClient.cpp]: funny mix of C and C++ styles. I'm fine with it.

File: swApiClient.cpp [hunk 5]
```diff
@@ -1,10 +1,27 @@
...
+  struct curl_slist *headers = NULL;
+  headers = curl_slist_append(headers, headerStr);
+  return headers;
+}
+
+static size_t WriteCallback(void *contents, size_t size, size_t nmemb, void *userp)
```
Comment (d33psky) [inline swApiClient.cpp]: maybe expand nmemb to number_of_elements or something to improve readability. same for userp

File: swApiClient.cpp (hunk 1 above)
Comment (d33psky) [reply swApiClient.cpp:58]: aye, let's keep this two lines as is. More clear.

File: swApiClient.cpp [hunk 6]
```diff
@@ -13,34 +30,80 @@ swApiClient Constructor
...
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postData);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
+    res = curl_easy_perform(curl);
+    if (res != 0) {
+      throw runtime_error(curl_easy_strerror(res));
+    }
+    curl_easy_cleanup(curl);
+    json j = json::parse(readBuffer);
+    this->accessToken = j["access_token"];
+  }
```
Comment (d33psky) [inline swApiClient.cpp]: else ...

File: swApiClient.cpp [hunk 7]
```diff
@@ -13,34 +30,80 @@ swApiClient Constructor
...
+    res = curl_easy_perform(curl);
+    if (res != 0) {
+      string error = curl_easy_strerror(res);
+      cout << error << endl;
+      cout << readBuffer << endl;
+      throw runtime_error(error);
+    }
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl);
+  }
```
Comment (d33psky) [inline swApiClient.cpp:120]: else ...

**Review by d33psky — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (mboylevt) [reply Makefile:18]: I FIXED IT

File: main.cpp (hunk 3 above)
Comment (mboylevt) [reply main.cpp]: it's consumer_key and consumer_secret: the .Authorize() function retrieves the bearer token.  I'll update that, and my styles :)

File: swApiClient.cpp (hunk 4 above)
Comment (mboylevt) [reply swApiClient.cpp]: It's 2018 Hans, I'm platform-fluid.

File: swApiClient.cpp (hunk 5 above)
Comment (mboylevt) [reply swApiClient.cpp]: yeah fair, it's a reference client: it should be readable

**Review by d33psky — APPROVED** [review]

**Pull request merged**